src/hooks/useOutsideClick.ts

```diff
@@ -0,0 +1,38 @@
+import { useEffect, useRef } from 'react';
+
+/**
+ * @description
+ * 이 훅은 지정한 요소의 외부를 클릭할 때 지정된 콜백 함수를 실행하도록 도와줍니다.
+ * 주로 모달 요소에서 외부 클릭을 감지하여 동작을 제어할 때 사용되며, useModal과 호환이 잘 되어 있습니다.
+ *
+ * @param callback 외부 클릭 시 호출할 콜백 함수
+ *
+ * @example
+ * const modalRef = useOutsideClick(() => {
+ *   console.log('외부 클릭');
+ * });
+ *
+ * return <div ref={modalRef}>모달 내용</div>;
+ */
+
+const useOutsideClick = (callback: () => void) => {
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    const onClickOutside = (e: MouseEvent) => {
+      if (ref.current && !ref.current.contains(e.target as Node)) {
+        callback();
+      }
+    };
+
+    document.addEventListener('mousedown', onClickOutside);
+
+    return () => {
+      document.removeEventListener('mousedown', onClickOutside);
+    };
+  }, [callback]);
+
+  return ref;
+};
```

Comment:
https://github.com/modern-agile-team/modern-kit/blob/main/packages/react/src/hooks/useOutsidePointerDown/index.ts

modern-kit의 `useOutsidePointerDown`도 참고해주셔도 좋아요.
useOutsidePointerDown은 데스크탑 환경이냐 모바일 환경이냐 따라 mouse이벤트를 바인딩 할지, touch 이벤트를 바인딩 할 지를 결정해줘요.

또한, touch와 mouse 이벤트를 모두 포함하는 광범위한 네이밍으로 pointer라는 네이밍을 사용했습니다.
아래는 Pointer Event에 대한 MDN의 내용을 가져왔습니다.

> Much of today's web content assumes the user's pointing device will be a mouse. However, since many devices support other types of pointing input devices, such as pen/stylus and touch surfaces, extensions to the existing pointing device event models are needed. [Pointer events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events#pointer_event) address that need.

https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events

Reply:
추가적으로 callback을 지금 함수로 받고있는데, 이거 사용처의 callback 내에서 state를 변경하면 무한 호출 될 수 있습니다 아래와 같은 예제입니다.

```
1. state가 변경
2. 컴포넌트 리렌더링
3. callback은 함수이며, 참조형이라 매번 재 생성 
4. useOutsideClick 내부 useEffect 의존성 배열에 걸림
```
```ts
const [value, setValue] = useState({ a: 1 });

useOutsideClick(() => {
  setValue({ ...value, b: 1 })
});
```

Reply:
저번에 이 훅에 대해서 @bluetree7878 님과 의논했던 기억이 있습니다. 코드를 짜던 중 구두로 이야기된 내용이라 히스토리는 남아있지 않지만 😢
@ssi02014 님이 제공해주신 레퍼런스대로  모던킷의 `usePreservedCallback` 훅을 이용해 callback 참조를 유지하는 내용과 동일하게 보완했던것으로 기억합니다!
다만. 아직 훅에 대해서 반영되지 않은 상태로 제가 해당 브랜치에 있는 훅을 가져오는 과정에서 그 부분이 빠진 것 같습니다.
히스토리가 남아있었으면 좋았을텐데 죄송합니다 ㅠㅠ

Reply:
@rhehfl 아하 이해했습니다 ~! 절대 죄송할건 아니에요 ㅋㅋ 히스토리 파악했습니다 👍

Reply:
모던킷에서[OutsideClick](https://github.com/modern-agile-team/modernkit/tree/main/packages/react/src/components/OutsideClick) 컴포넌트와 `useOutsidePointerDown`라는 훅도 존재했었네요 ㅎㅎ 처음에 `OutsideClick` 컴포넌트를 보고 `useOutsidePointerDown`같은 훅이 없는 줄 알았습니다!
구현도 `useOutsidePointerDown`훅을 사용하는게 더 좋을 것 같네요!
훅 구현해주신 @bluetree7878님은 어떻게 생각하시나용??

Reply:
@rhehfl 참고로, OutsideClick은 useOutsidePointerDown을 선언적으로 활용 할 수 있는 컴포넌트이지만, 해당 컴포넌트는 개선이 필요한 상황입니다 현재 진행중에 있습니다. 네이밍 변경 + 내부 로직 수정

Reply:
**Login.tsx**
```tsx
interface LoginProps {
  openModal: () => void;
  closeModal: () => void;
}

export default function Login({ closeModal }: LoginProps) {
  const modalRef = useOutsideClick(closeModal);
   ```
   맞습니다. 그때 @rhehfl 님과 함께 `usePreservedCallback `훅을 이용해서 쓸까말까에 대한 의논을 했었는데요. 현재 `Login`컴포넌트 쪽에서 쓰고 있는데 현재 이 상태에서 확인해봤을 때에는 `usePreservedCallback`훅을 쓰든 안 쓰든 똑같은 상태여서 오히려 넣었을 때 '성능이슈가 있지 않을까' 하고 최종적으로 일단 빼기로 했었습니다.
 
   @rhehfl @ssi02014 저 또한 모던킷에서 `useOutsidePointerDown` 이러한 훅이 있었는지 몰랐네요.. 이 훅을 사용해보는 것도 괜찮을 것 같습니다 ㅎㅎ

Reply:
@bluetree7878 현재 사용에서는 크게 이슈가 없을 수도 있습니다!! ㅎㅎ 제가 위에서 언급드린 케이스가 만약에 있다면, 이 부분에서 이슈가 발생할거에요 `공통 훅`이라는 관점에서 여러 케이스를 대응할 수 있으면 좋을 것 같습니다.

Reply:
확실히 공통 훅이라는 게 결국 여러 케이스에 사용하려고 쓰는 것이니 `usePreservedCallback`을 이용하여 callback 참조를 유지시키도록 하는 게 좋을 것 같습니다 감사합니다!